onyo/lib/tests/test_git.py

```diff
@@ -0,0 +1,345 @@
+    to existing repositories.
+    """
+    # initialize the temp_path as an git repository
+    subprocess.run(['git', 'init', tmp_path])
+
+    # test that `GitRepo()` instantiates (and sets the root to) the object
+    # correctly
+    new_repo = GitRepo(tmp_path)
+    assert new_repo.root.samefile(tmp_path)
+
+    # create a sub-directory to test the find-root behavior
+    subprocess.run(['mkdir', '-p', tmp_path / "sub-directory"])
+
+    # with `find_root=False` it excepts other existing paths e.g. sub-dirs
+    new_repo = GitRepo(tmp_path / "sub-directory", find_root=False)
+    assert new_repo.root.samefile(tmp_path / "sub-directory")
+
+    # with `find_root=True` it must find the root and set it appropriately
+    new_repo = GitRepo(tmp_path / "sub-directory", find_root=True)
+    assert new_repo.root.samefile(tmp_path)
+
+
+def test_GitRepo_find_root(tmp_path: Path) -> None:
+    """
+    `GitRepo.find_root()` MUST identify the root of a repository for the root
+    itself and sub-directories of a repository.
+    If called on a Path which is not a git repository it must raise an
+    `OnyoInvalidRepoError`.
+    """
+
+    # test GitRepo.find_root() with an existing but non-repo path raises the
+    # expected OnyoInvalidRepoError
+    with pytest.raises(OnyoInvalidRepoError):
+        GitRepo.find_root(tmp_path)
+
+    # Path GitRepo.find_root() with a Path that does not exist at all raises the
+    # correct OnyoInvalidRepoError
+    with pytest.raises(OnyoInvalidRepoError):
+        GitRepo.find_root(Path("This/path/does/not/exist"))
+
+    # initialize a new git repo to test correct behavior when git exists
+    subprocess.run(['git', 'init', tmp_path])
+    subprocess.run(['mkdir', '-p', tmp_path / 'existing/directory'])
+
+    # test that `GitRepo.find_root(root)` returns the same root path
+    assert GitRepo.find_root(tmp_path).samefile(tmp_path)
+
+    # test that `GitRepo.find_root()` returns correct root for a sub-directory
+    assert GitRepo.find_root(tmp_path / 'existing/directory').samefile(tmp_path)
+
+    # test that `GitRepo.find_root()` raises the correct error when called on
+    # a non-existing sub-path under an existing root
+    with pytest.raises(OnyoInvalidRepoError):
+        GitRepo.find_root(tmp_path / 'non-existing/directory')
+
+
+def test_GitRepo_restore_staged(tmp_path: Path) -> None:
+    """
+    `GitRepo.restore_staged()` must restore all staged files in the repository.
+    If no files are staged, it should not raise an error.
+    If there are modified or untracked files, they should not be changed.
+    """
+    subprocess.run(['git', 'init', tmp_path])
+    new_git = GitRepo(tmp_path)
+    test_file = new_git.root / 'asset_for_test.0'
+    test_file.touch()
+    new_git.stage_and_commit(test_file, message="Create file for test")
+
+    # test that no error is raised, when called on a clean repository
+    new_git.restore_staged()
+
```

Comment:
There's no assertion on what `restore_staged` did or didn't do, just that it doesn't raise. I think minimum is an assertion that the repo is still clean. One could also test, that it's not only clean but still on the same commit.
There are similar cases in other tests, where you call the to-be-tested function and have no assertions on the result.